Added more information to Gamepad IPC crashes

abigailfox opened the pull request:
#### fe27d21981f597d3ab92dc16344b49343d183754
<pre>
Added more information to Gamepad IPC crashes
<a href="https://bugs.webkit.org/show_bug.cgi?id=247468">https://bugs.webkit.org/show_bug.cgi?id=247468</a>
rdar://101812342

Reviewed by David Kilzer.

* Source/WebKit/WebProcess/Gamepad/WebGamepadProvider.cpp:
(WebKit::WebGamepadProvider::setInitialGamepads):
(WebKit::WebGamepadProvider::gamepadConnected):
(WebKit::WebGamepadProvider::gamepadDisconnected):

Canonical link: <a href="https://commits.webkit.org/256344@main">https://commits.webkit.org/256344@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a500bd2312a01f0be0f9fa7935fdcb2f1b025872

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/95480 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/4749 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/28534 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/105062 "Built successfully") | 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/4776 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/33482 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/87850 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/100925 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/101144 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/3489 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/82092 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/30562 "Passed tests") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/85381 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/87297 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/73401 "Found 2 new API test failures: /WebKitGTK/TestWebKitAccessibility:/webkit/WebKitAccessibility/accessible/state-changed, /WebKitGTK/TestWebKitAccessibility:/webkit/WebKitAccessibility/accessible/event-listener (failure)") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/66/builds/39248 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/18852 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/67/builds/36949 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/20150 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/4385 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/40943 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/42803 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/42932 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/39397 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

ddkilzer commented on Source/WebKit/WebProcess/Gamepad/WebGamepadProvider.cpp:
```diff
@@ -39,6 +39,13 @@
 using namespace WebCore;
 
+#define WP_MESSAGE_CHECK(assertion, ...) { \
+    if (UNLIKELY(!assertion)) { \
```
This needs extra parenthesis around `assertion` (even though you include them below) to be defensive:
```
    if (UNLIKELY(!(assertion))) { \
```

ddkilzer commented on Source/WebKit/WebProcess/Gamepad/WebGamepadProvider.cpp:
```diff
@@ -76,8 +83,8 @@ void WebGamepadProvider::gamepadConnected(const GamepadData& gamepadData, EventM
-
-    ASSERT(!m_gamepads[gamepadData.index()]);
+    
+    WP_MESSAGE_CHECK((!m_gamepads[gamepadData.index()]), gamepadData.index(), m_gamepads.size());
```
Nit:  Might be worth saving the old size above:
```
    auto oldGamepadsSize = m_gamepads.size();
    ...
    WP_MESSAGE_CHECK((!m_gamepads[gamepadData.index()]), oldGamepadsSize, gamepadData.index(), m_gamepads.size());
```

ddkilzer commented on Source/WebKit/WebProcess/Gamepad/WebGamepadProvider.cpp:
```diff
@@ -76,8 +83,8 @@ void WebGamepadProvider::gamepadConnected(const GamepadData& gamepadData, EventM
     }
-
-    ASSERT(!m_gamepads[gamepadData.index()]);
+    
```
Nit:  No need to add this whitespace on Line 86 (which doesn't show up here, but shows in a prior comment).

ddkilzer reviewed (changes requested):
r- to add parentheses to `assertion` in `WP_MESSAGE_CHECK()`.

ddkilzer reviewed (approved):
r=me

abigailfox committed:
Added more information to Gamepad IPC crashes
https://bugs.webkit.org/show_bug.cgi?id=247468
rdar://101812342

Reviewed by David Kilzer.

* Source/WebKit/WebProcess/Gamepad/WebGamepadProvider.cpp:
(WebKit::WebGamepadProvider::setInitialGamepads):
(WebKit::WebGamepadProvider::gamepadConnected):
(WebKit::WebGamepadProvider::gamepadDisconnected):

Canonical link: https://commits.webkit.org/256344@main
